File: docs/manual/java/guide/services/code/docs/services/AdditionalRouters.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.services;

import akka.NotUsed;
import play.api.mvc.Handler;
import play.api.mvc.RequestHeader;
import play.api.routing.Router;
import play.api.routing.SimpleRouter;
import play.mvc.Http;
import play.routing.RoutingDsl;
import scala.PartialFunction;

import javax.inject.Inject;
import java.io.File;
import java.util.List;
import static play.mvc.Results.ok;

import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.Service;
import com.lightbend.lagom.javadsl.api.ServiceCall;

import static com.lightbend.lagom.javadsl.api.Service.named;
import static com.lightbend.lagom.javadsl.api.Service.pathCall;
import static com.lightbend.lagom.javadsl.api.ServiceAcl.path;
import static java.util.concurrent.CompletableFuture.completedFuture;
import com.google.inject.AbstractModule;
import com.lightbend.lagom.javadsl.server.ServiceGuiceSupport;
import static java.util.stream.Collectors.joining;

public class AdditionalRouters {

  // #hello-service
  interface HelloWordService extends Service {

    ServiceCall<NotUsed, String> hello(String name);

    @Override
    default Descriptor descriptor() {
      return named("fileupload")
          .withCalls(pathCall("/api/hello/:name", this::hello))
          .withAutoAcl(true)
          .withServiceAcls(path("/api/files"));
    }
  }
  // #hello-service

  public static class HelloWordServiceImpl implements HelloWordService {
    @Override
    public ServiceCall<NotUsed, String> hello(String name) {
      return input -> completedFuture("Hello, " + name);
    }
  }

  // ------------------------------------------------------------------------
  // SomePlay Router example
  class SomePlayRouterExample {

    class SomePlayRouter implements SimpleRouter {
      @Override
      public PartialFunction<RequestHeader, Handler> routes() {
        throw new RuntimeException("Not implemented");
      }
    }

    public class HelloWorldModuleDI {
      // example to show API that supports DI (Guice)
      // #lagom-module-some-play-router-DI
      public class HelloWorldModule extends AbstractModule implements ServiceGuiceSupport {
        @Override
        protected void configure() {
          bindService(
              HelloService.class, HelloServiceImpl.class, additionalRouter(SomePlayRouter.class));
        }
      }
      // #lagom-module-some-play-router-DI
    }

    public class HelloWorldModuleInstance {
      // example to show API that supports passing instance
      // #lagom-module-some-play-router-instance
      public class HelloWorldModule extends AbstractModule implements ServiceGuiceSupport {
        @Override
        protected void configure() {
          bindService(
              HelloService.class, HelloServiceImpl.class, additionalRouter(new SomePlayRouter()));
        }
      }
      // #lagom-module-some-play-router-instance
    }
  }

  // ------------------------------------------------------------------------
  class FileUploadRouterExample {
    // #file-upload-router
    class FileUploadRouter implements SimpleRouter {

      private final Router delegate;

      @Inject
      public FileUploadRouter(RoutingDsl routingDsl) {
        this.delegate =
            routingDsl
                .POST("/api/files")
                .routingTo(
                    request -> {
                      Http.MultipartFormData<File> body = request.body().asMultipartFormData();
                      String response =
                          body.getFiles().stream()
                              .map(f -> f.getRef().getAbsolutePath())
                              .collect(joining(",", "Uploaded[", "]"));
                      return ok(response);
                    })
                .build()
                .asScala();
      }

      @Override
      public PartialFunction<RequestHeader, Handler> routes() {
        return delegate.routes();
      }
    }
    // #file-upload-router

    // #lagom-module-file-upload
    public class HelloWorldModule extends AbstractModule implements ServiceGuiceSupport {
      @Override
      protected void configure() {
        bindService(
            HelloService.class, HelloServiceImpl.class, additionalRouter(FileUploadRouter.class));
      }
    }
    // #lagom-module-file-upload
  }
}

File: docs/manual/scala/guide/services/code/AdditionalRouters.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl.services

import akka.NotUsed
import com.lightbend.lagom.scaladsl.api.Service
import com.lightbend.lagom.scaladsl.api.ServiceAcl
import com.lightbend.lagom.scaladsl.api.ServiceCall
import play.api.routing.SimpleRouter
import scala.concurrent.Future

package additionalrouters {

  package helloservice {

    //#hello-service
    trait HelloService extends Service {

      def hello(id: String): ServiceCall[NotUsed, String]

      final override def descriptor = {
        import Service._
        named("hello")
          .withCalls(
            pathCall("/api/hello/:id", hello _).withAutoAcl(true)
          )
          .withAcls(
            // extra ACL to expose additional router endpoint on ServiceGateway
            ServiceAcl(pathRegex = Some("/api/files"))
          )
      }
    }
    //#hello-service

    //#hello-service-impl
    class HelloServiceImpl extends HelloService {
      override def hello(name: String) = ServiceCall { _ =>
        Future.successful(s"Hello $name!")
      }
    }
    //#hello-service-impl
  }

  package lagomapplication {

    import helloservice._

    import com.lightbend.lagom.scaladsl.server._
    import play.api.libs.ws.ahc.AhcWSComponents
    import com.softwaremill.macwire._

    package sompeplayrouter {

      import play.api.routing.Router.Routes

      class SomePlayRouter extends SimpleRouter {
        override def routes: Routes = ???
      }

      abstract class HelloApplication(context: LagomApplicationContext)
          extends LagomApplication(context)
          with AhcWSComponents {

        //#lagom-application-some-play-router
        override lazy val lagomServer =
          serverFor[HelloService](wire[HelloServiceImpl])
            .additionalRouter(wire[SomePlayRouter])
        //#lagom-application-some-play-router
      }
    }

    package fileuploadrouter {
      //#file-upload-router
      import play.api.mvc.DefaultActionBuilder
      import play.api.mvc.PlayBodyParsers
      import play.api.mvc.Results
      import play.api.routing.Router
      import play.api.routing.sird._

      class FileUploadRouter(action: DefaultActionBuilder, parser: PlayBodyParsers) {
        val router = Router.from {
          case POST(p"/api/files") =>
            action(parser.multipartFormData) { request =>
              val filePaths = request.body.files.map(_.ref.getAbsolutePath)
              Results.Ok(filePaths.mkString("Uploaded[", ", ", "]"))
            }
        }
      }
      //#file-upload-router

      abstract class HelloApplication(context: LagomApplicationContext)
          extends LagomApplication(context)
          with AhcWSComponents {

        //#lagom-application-file-upload
        override lazy val lagomServer =
          serverFor[HelloService](wire[HelloServiceImpl])
            .additionalRouter(wire[FileUploadRouter].router)
        //#lagom-application-file-upload
      }
    }

  }
}
